crates/sparrow-materialize/src/materialize/materialization_control.rs

```diff
@@ -0,0 +1,219 @@
+                match handle
+                    .await
+                    .into_report()
+                    .change_context(Error::Internal)?
+                    .change_context(Error::Materialization)
+                {
+                    Ok(_) => {
+                        self.status =
+                            MaterializationProgress::Stopped(ProgressInformation::default());
+                    }
+                    Err(e) => {
+                        self.status =
+                            MaterializationProgress::Failed(ProgressInformation::default(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Running(progress) => {
+                match handle
+                    .await
+                    .into_report()
+                    .change_context(Error::Internal)?
+                    .change_context(Error::Materialization)
+                {
+                    Ok(_) => {
+                        self.status = MaterializationProgress::Stopped(progress.clone());
+                    }
+                    Err(e) => {
+                        self.status = MaterializationProgress::Failed(progress.clone(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Stopped(_) => {
+                tracing::info!("cannot stop; materialization already stopped")
+            }
+            MaterializationProgress::Failed(_, _) => {
+                tracing::info!("cannot stop; materialization already failed")
+            }
+        };
+
+        Ok(())
+    }
+
+    /// Gets the status of the materialization.
+    ///
+    /// Note that progress is reported periodically, and may not contain the exact
+    /// up-to-date numbers.
+    pub fn get_status(&self) -> MaterializationStatus {
+        let progress = self.progress_receiver.borrow();
+        match &*progress {
+            MaterializationProgress::Uninitialized => MaterializationStatus {
+                id: self.id.clone(),
+                progress: ProgressInformation::default(),
+                state: State::Uninitialized,
+                error: None,
+            },
+            MaterializationProgress::Running(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Running,
+                error: None,
+            },
+            MaterializationProgress::Stopped(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Stopped,
+                error: None,
+            },
+            MaterializationProgress::Failed(progress, e) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Failed,
+                error: Some(e.to_string()),
+            },
+        }
+    }
+}
```

Comment:
Thought experiment -- may be easier / cleaner to clone the progress info and then divide it up to make the result.

```suggestion
    pub fn get_status(&self) -> MaterializationStatus {
        // Review note: While excess cloning can be a concern, it doesn't seem too bad here.
        // We need to make a copy of the progress to return anyway.
        let progress = self.progress_receiver.borrow().clone();
        let error = match progress {
          MaterializationProgress::Failed(_, e) => Some(e.to_string()),
          _ => None
        };
 
        let progress = match progress {
          MaterializationProgress::Uninitialized => ProgressInformation::default(),
          MaterializationProgress::Running(progress) => progress,
          MaterializationProgress::Stopped(proggres) => progress,
          MaterializationProgress::Failed(progress, _) => progress,
        };
        MaterializationStatus {
          id: self.id.clone(),
          progress,
          state: ???,
          error,
        }
    }
}
```